spec.md

```diff
@@ -224,7 +224,10 @@ help intermediate gateways determine how to route the events.
 * Constraints:
   * REQUIRED
   * MUST be a non-empty string
-  * MUST be unique within the scope of the producer
+  * MUST be unique within the scope of the producer. In other words:
+    events with identical "scope" and "id" values MUST be duplicates
```

Comment:
Did you mean `source`, not `scope`?

Anyway, I agree with https://github.com/cloudevents/spec/issues/331#issuecomment-463754907 - scope should be both `source` and `type`.

E.g. `com.github.pull.create` and `/cloudevents/spec/pull/123` together form the scope of the producer.

Reply:
WIll fix source/scope typo.

I prefer that source+id be unique and type be orthogonal. 

A source defines exactly the scope within which ids are unique. If you need ids to be unique by type, then you can create a source per type. If you don't, then a single source can emit events of many types. In either case the consumers need only consider source+id for de-duplication.

Reply:
not sure source+type is always unique...  `/cloudevents/spec/pull/123`+`com.github.com.comment` is not unique. Wouldn't ID need to be in there some place?

Reply:
I agree: source+id is unique. Nothing else should be required/assumed to be unique. type is a separate issue and best not conflated with deduplication. If you want to conflate sources and types you can (e.g. include type info in the source URI and only emit events of that type) but you don't have to. Consumers should not be required to include type in the calculation for de-duplication.

Reply:
@duglin Yes, the ID needs to be in there! I only meant to say source+type=scope, and scope+id is unique.

Reply:
> Consumers should not be required to include type in the calculation for de-duplication.

@alanconway You apparently feel strongly about this. Why? Whether I have to use two or three fields for de-duplication doesn't seem to make a huge difference to me.

Reply:
I argue that id and type should be separate both for simplicity and clear semantics

simplicity:
- 2 fields is better than 3 unless there's value in including the 3rd. What's the value?
- when efficiency is a concern you can generate short IDs, but you can't control the length of type names - which may be an orders of magnitude longer.

semantics:
- "id" is for de-duplication, it has no other reason to exist. You also need "source" if there are many sources, because separate sources can't co-operate to generate unique ids.  You don't need anything else.
- "type" is for classifying events for processing. That has nothing at all to do with deduplication. Sources can produce events of one type or many.

I've seen type and id conflated before, I've done it myself :) The early thinking is "it'll be easier for producers to generate unique IDs per type." The reality that hits later is that it is harder for consumers to use IDs that have to be coupled with a type to mean anything. 

Consumers tend to naturally collect messages by source, then they just want a simple hashable ID to check duplicates. If you include the type you force a  bunch of awkward code around:

    type UsableID struct { Type, ID string }

Depending on the language you might have to define hash functions or comparisons to use these as map keys etc  (you can't just concatenate type and id into a single string because that might clash with some other type-id combo, unless you know something about the syntax of type and ID.) It's all just a pain.

Conflating type with ID is certainly implementable, but IMO makes implementation harder and less efficient.

Reply:
If the ID is semantically meaningful (for example, a TrueTime timestamp or a vector clock), then middleware may need to preserve the ID when transforming an event's `type`. In this case, `source` + `id` may need to be non-unique, or we could guide that the middleware should set embed the originating `id` into a longer `id` related to the source in such a way that the recipients know how to extract the originating `id`.

Reply:
resolving, please check latest and re-raise any issues.